Fix editorial nits (#964)
dotnet/csharpstandard: standard/lexical-structure.md

>
> *end example*

> A *relational_expression* ([§12.12.1](expressions.md#12121-general)) can have the form “*relational_expression* `is` *type*” or “*relational_expression* `is` *constant_pattern*,” either of which might be a valid parse of a qualified identifier. In this case, an attempt is made to bind it as a type ([§7.8.1](basic-concepts.md#781-general)); however, if that fails, it is bound as an expression, and the result is a constant.

Reviewer: "the result is a constant" sounds like a classification, whereas we're stating a constraint on the code. (I think this part of me finding "shall be" generally more confusing than "must", so we may be able to ignore this as just a bit of Jon confusion, but I'd be interested to hear other views...)

Reviewer 2: The language is wrong, e.g. “an attempt is made” is an operational description of the implementation of a specification rather than a specification. What it should say is that if both forms are valid (syntactically & semantically) parses of the input, i.e. there is an ambiguity, then the former (`is` *type*) shall be selected. There are similar situations elsewhere and the wording used might be suitable here, but no wordsmithing from me tonight – its too late!

Reviewer 3: Here's my first attempt at a wordsmithing.  Rex may have better ideas:

```suggestion
A *relational_expression* ([§12.12.1](expressions.md#12121-general)) can have the form “*relational_expression* `is` *type*” or “*relational_expression* `is` *constant_pattern*,” either of which might be a valid parse of a qualified identifier. When both forms result in correct code, the binding to a type ([§7.8.1](basic-concepts.md#781-general)) is preferred. If that fails, it is bound as an expression, and the result is a constant.
```

Reviewer: Updated "bindint" to "binding" in the suggestion.

Reviewer: I'm not sure about "correct code". It's more *valid* than *correct*, but I'm not sure whether that would be right either. Legal code? Or maybe talk about name lookups succeeding?

Reviewer 2: First, I think this should probably be in $12.12.1 and not here – or in both, I've added #965 to address that and will suggest changes here for now.

It is not unusual for alternatives to overlap in language grammars, sometimes just because it produces a shorter/neater grammar, sometimes as some semantic information is required to resolve the syntactic choice, etc. For example in $12.19.1 we have:

> When recognising an *anonymous_function_body* if both the *null_conditional_invocation_expression* and *expression* alternatives are applicable then the former shall be chosen.

We can resolve the issue here by following the same pattern:

```suggestion
When recognising a *relational_expression* ([§12.12.1](expressions.md#12121-general)) if both the “*relational_expression* `is` *type*” and “*relational_expression* `is` *constant_pattern*” alternatives are applicable, and *type* resolves to an accessible type, then the “*relational_expression* `is` *type*” alternative shall be chosen.
```

I would refrain from explaining the overlap between the two alternatives further, the grammar does that, but if that is desired it should be added as a *Note*.

The bit “and the result is a constant” in the original is both confusing (is it referring to the *constant_pattern* or the result of the `is`?) and unneeded – any constness comes from the semantics of *constant_pattern* and not the `is` operator whose result is not a constant in general!)
